File: src/main/java/com/theonlytails/ruby/items/Ruby.java
package com.theonlytails.ruby.items;

import com.theonlytails.ruby.TheOnlyTails;
import com.theonlytails.ruby.entities.RubySheepEntity;
import com.theonlytails.ruby.init.RubyEntityTypes;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.SpawnReason;
import net.minecraft.entity.passive.SheepEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ActionResultType;
import net.minecraft.util.Hand;
import org.jetbrains.annotations.NotNull;

public class Ruby extends Item {
    public Ruby() {
        super(new Item.Properties().group(TheOnlyTails.RUBY));
    }

    @Override
    public @NotNull ActionResultType itemInteractionForEntity(@NotNull ItemStack stack,
                                                              @NotNull PlayerEntity playerIn,
                                                              @NotNull LivingEntity target,
                                                              @NotNull Hand hand) {
        if (target instanceof SheepEntity) {
            SheepEntity sheepentity = (SheepEntity)target;
            if (sheepentity.isAlive() && !sheepentity.getSheared() && !(sheepentity instanceof RubySheepEntity)) {
                if (!playerIn.world.isRemote) {
                    RubySheepEntity rubySheepEntity = RubyEntityTypes.RUBY_SHEEP.get().create(playerIn.world);
                    if (rubySheepEntity != null) {
                        rubySheepEntity.setLocationAndAngles(sheepentity.getPosX(), sheepentity.getPosY(), sheepentity.getPosZ(), sheepentity.rotationYaw, sheepentity.rotationPitch);
                        rubySheepEntity.onInitialSpawn(playerIn.world, playerIn.world.getDifficultyForLocation(rubySheepEntity.getPosition()), SpawnReason.CONVERSION, null, null);
                        playerIn.world.addEntity(rubySheepEntity);
                        sheepentity.remove();
                    }

                }

                return ActionResultType.func_233537_a_(playerIn.world.isRemote);
            }
        }

        return ActionResultType.PASS;
    }
}

File: src/main/kotlin/com/theonlytails/rubymod/items/Ruby.kt
package com.theonlytails.rubymod.items

import com.theonlytails.rubymod.entities.RubySheepEntity
import com.theonlytails.rubymod.registries.EntityTypeRegistry
import com.theonlytails.rubymod.rubyTabProperty
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.SpawnReason
import net.minecraft.entity.passive.SheepEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.util.ActionResultType
import net.minecraft.util.Hand
import net.minecraft.world.IServerWorld

/**
 * Holds the custom functionality of rubies.
 *
 * @author TheOnlyTails
 */
class Ruby : Item(rubyTabProperty) {
	override fun interactLivingEntity(
		stack: ItemStack,
		playerIn: PlayerEntity,
		target: LivingEntity,
		hand: Hand,
	): ActionResultType =
		if (target is SheepEntity && target.isAlive && !target.isSheared && target !is RubySheepEntity) {
			if (!playerIn.level.isClientSide) {
				EntityTypeRegistry.rubySheep.create(playerIn.level)?.apply {
					moveTo(
						target.x,
						target.y,
						target.z,
						target.yRot,
						target.xRot
					)

					finalizeSpawn(
						playerIn.level as IServerWorld,
						playerIn.level.getCurrentDifficultyAt(blockPosition()),
						SpawnReason.CONVERSION,
						null,
						null
					)

					playerIn.level.addFreshEntity(this)
					target.remove()

					stack.shrink(1)
				}
			}

			ActionResultType.sidedSuccess(playerIn.level.isClientSide)
		} else ActionResultType.PASS
}